Chat data + prompt template tutorial (#823)
pytorch/torchtune: docs/source/tutorials/chat.rst

=================================
Fine-tuning LLaMA3 with chat data
=================================

In this tutorial, we'll demystify what prompt templates are and when you'll need them
and talk about the differences between prompt templating for LLaMA2 and LLaMA3. Then,

> we'll wrap up with a LLaMA3 finetuning example on a custom chat dataset.

Reviewer: To me this doesn't actually fully line up with the title of the tutorial. Also kind of a nit but I wouldn't use "demystify" in the opening sentence, I think it can potentially turn people off by making a topic seem too complex.

Author: Maybe frame it more as, "llama3 handles chat formats differently, here's what you need to know"? Also what would you suggest for the title?

.. grid:: 2

    .. grid-item-card:: :octicon:`mortar-board;1em;` You will learn:

      * What is a prompt template and how is it different from special tokens
      * When to use and not to use a prompt template

>       * How to use your own chat dataset to fine-tune LLaMA3

Reviewer: Why not just lead with this one, I feel like the other points are too in the weeds and kinda unclear tbh

Author: How about:

- Fine tune chat data with llama3
- what prompt templates are and how to use them

    .. grid-item-card:: :octicon:`list-unordered;1em;` Prerequisites

>       * Be familiar with :ref:`configuring datasets<dataset_tutorial_label>`

Reviewer: I'm not super on the difference between this tutorial and the existing one. Could we split into something like beginner and advanced tutorials? Beginner being low degree of customization, advanced being "here's everything you need to know to set up your own dataset from scratch". Or alternatively make this purely Llama3-focused and go through all those nuances

Author: Yeah I feel like there's two main concepts I want to convey in this tutorial: dealing with llama3's new formatting and finetuning on a custom dataset with llama3. I'm thinking of keeping those two in this tutorial for now and see how to dedupe with the dataset tutorial later, do you think some topics should move to the other tutorial?

Author: I also think a bit of redundancy is ok or even good, it's helpful to see similar information conveyed through different problems, and this llama3 data tutorial and dataset tutorials approach the concepts different

      * Know how to :ref:`download LLaMA3 weights <llama3_label>`

Prompt templates & special tokens
---------------------------------

Prompt templates and tokenization schemes are often conflated, and it's sometimes

> unclear what you need to do to format your data to optimize the training performance

Reviewer: Imo this is not a good starting point. You need to set the stage of the tutorial with "here's a specific problem we're going to solve that you care about"

of your model. Let's walk through the LLaMA2/LLaMA3 templates to better understand

> the distinction.

Reviewer: This is a better motivation for a section: learning how Llama3 format differs from Llama2 format

In torchtune, a prompt template is simply a structured way to format your prompt
into the model. The prompt template adds flavor text to prime the model to expect
a certain structure when it is used for inference. From the `official LLaMA2 prompt
template guide <https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-2>`_
for the LLaMA2 chat model, we can see that special tags are added:

.. code-block:: text

    <s>[INST] <<SYS>>
    {{ system_prompt }}
    <</SYS>>

    {{ user_message_1 }} [/INST] {{ model_answer_1 }} </s>

>     <s>[INST] {{ user_message_2 }} [/INST]

Reviewer: This is nice but you need to give actual examples. Take a specific input text and walk through all the steps: prompt format, tokenization, special tokens, etc. Give a consistent simple example that readers can always anchor back to for their understanding.

Since the chat model was pretrained with this prompt template, if you want to run
inference on the model, you'll need to use the same template for optimal performance
on chat data. Otherwise, the model will just perform standard text completion, which
may or may not align with your intended use case.

If you look at our :class:`~torchtune.data.Llama2ChatFormat` class, you'll notice that
we don't include the :code:`<s>` and :code:`</s>` tokens. These are the beginning-of-sequence
(BOS) and end-of-sequence (EOS) tokens that are represented differently in the tokenizer
than the rest of the prompt template. Let's tokenize some examples with the
:class:`~torchtune.modules.tokenizers.SentencePieceTokenizer` used by LLaMA2 to see
why.

> .. code-block:: python

Reviewer: The whole remainder of this section is nice and not really related to formatting. This should be its own section on tokenization and special tokens imo

    from torchtune.modules.tokenizers import SentencePieceTokenizer

    tokenizer = SentencePieceTokenizer("/tmp/Llama-2-7b-hf/tokenizer.model")
    text = "[INST] <<SYS>> Hello"
    tokens = tokenizer.encode(text, add_bos=True, add_eos=True)
    print(tokens)
    # [1, 518, 25580, 29962, 3532, 14816, 29903, 6778, 15043, 2]

We've added the BOS and EOS tokens when encoding our example text. This shows up
as IDs 1 and 2. We can verify that these are our BOS and EOS tokens.

.. code-block:: python

    print(tokenizer.spm_model.piece_to_id("<s>"))
    # 1
    print(tokenizer.spm_model.piece_to_id("</s>"))
    # 2

The BOS and EOS tokens are what we call special tokens, because they have their own
reserved token IDs. This means that they will index to their own individual vectors in
the model's learnt embedding table. The rest of the prompt template tags, :code:`[INST]`

> and :code:`<<SYS>>` are tokenized as normal text and not their own IDs.

Reviewer: Is the point here that `[INST]` and `<<SYS>>` are not special tokens? If so, just say that (though tbh I wouldn't even go into it, and would just make this section about tokenization and special tokens)

Author: Yes, but I would like to show that they are tokenized differently to illustrate my point, so I need to talk about it in the context of tolenization

    # [529, 29879, 29958]

When should I use a prompt template?
------------------------------------

When to use or not to use a prompt template is governed by what your desired inference
behavior is. You should use a prompt template in the following scenarios:

1. You are running inference on the base model and it was pre-trained with a prompt
template
2. You want to prime a fine-tuned model to expect a certain prompt structure on inference
for a specific task

It is not strictly necessary to fine-tune with a prompt template, but generally you
want the model to perform some sort of task, which will require some formatting of
the prompt.

For example, the :class:`~torchtune.data.SummarizeTemplate` provides a lightweight
structure to prime your fine-tuned model for prompts asking to summarize text.

.. code-block:: python

    f"Summarize this dialogue:\n{dialogue}\n---\nSummary:\n"

You can fine-tune LLaMA2 with this template even though the model was originally pre-trained
with the :class:`~torchtune.data.Llama2ChatFormat`, as long as this is what the model
sees during inference. The model should be robust enough to adapt to a new template.

> Special tokens in LLaMA3

Reviewer: These sections don't flow together super clearly. Prompt template + special tokens section, followed by when to use prompt template, followed by specific handling of special tokens for Llama3. Why not just tackle prompt template and special tokens as their own standalone entities and walk through them fully one at a time?

Author: Yeah, let me see if I can separate these two a bit better.

------------------------

LLaMA3 `overhauled <https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-3>`_
the special tokens and prompt templating from LLaMA2 to better support multiturn conversations.

.. code-block:: text

>     <|begin_of_text|><|start_header_id|>system<|end_header_id|>

Reviewer: Again, give a specific example here

    {{ system_prompt }}<|eot_id|><|start_header_id|>user<|end_header_id|>

    {{ user_message_1 }}<|eot_id|><|start_header_id|>assistant<|end_header_id|>

    {{ model_answer_1 }}<|eot_id|><|start_header_id|>user<|end_header_id|>

Now, all the additional tags are special tokens, and no prompt template is required.
You can see how they get encoded as their own token IDs below:

.. code-block:: python

    from torchtune.modules.tokenizers import TikTokenTokenizer

    tokenizer = TikTokenTokenizer("/tmp/Meta-Llama-3-8B/original/tokenizer.model")
    print(tokenizer._encode_special_token("<|begin_of_text|>"))
    # 128000
    print(tokenizer._encode_special_token("<|eot_id|>"))

>     # 128009

Reviewer: Could be worth a quick digression into the configurations for TikToken encoding for allowed and disallowed special tokens and how we will treat any special tokens appearing in the text just as regular text (idk if this is too in the weeds though)

Author: I think that's way too deep imo, and no one should be playing with those APIs

The best part is - all these special tokens are handled purely by the tokenizer.
That means you won't have to worry about messing up any required prompt templates!

Fine-tuning on a custom chat dataset
------------------------------------

Let's test our understanding by trying to fine-tune the LLaMA3-8B model with a custom
chat dataset. We'll walk through how to set up our data so that it can be tokenized

> correctly and fed into our model.

Reviewer: 🍿

You can see how they get encoded as their own token IDs below:

.. code-block:: python

    from torchtune.modules.tokenizers import TikTokenTokenizer

    tokenizer = TikTokenTokenizer("/tmp/Meta-Llama-3-8B/original/tokenizer.model")
    print(tokenizer._encode_special_token("<|begin_of_text|>"))
    # 128000
    print(tokenizer._encode_special_token("<|eot_id|>"))
    # 128009

The best part is - all these special tokens are handled purely by the tokenizer.
That means you won't have to worry about messing up any required prompt templates!

Fine-tuning on a custom chat dataset
------------------------------------

Let's test our understanding by trying to fine-tune the LLaMA3-8B model with a custom
chat dataset. We'll walk through how to set up our data so that it can be tokenized
correctly and fed into our model.

We'll be using the `LIMA <https://arxiv.org/pdf/2305.11206.pdf>`_ dataset, a small,
carefully crafted dataset from Q&A selected for responses in the style of a helpful
AI assistant. Looking at the :class:`~torchtune.datasets.ChatDataset` class, you'll
notice several parameters you'll need to provide to configure it:

.. code-block:: python

>     class ChatDataset(Dataset):

Reviewer: Overall I like this section. Still a couple of comments: (1) I think it can be a bit overwhelming to just provide the general `ChatDataset` API from the get-go without a ton of context. Why not start with a single sample, show the format we want to get it into for tokenization, then walk through each of the necessary steps to get there? Then use that to motivate the exact definition of lima_dataset given below.

Author: That's a sound approach

    # '<<SYS>>'

It's important to note that you should not place the special reserved tokens in your
input prompts manually, as it will be treated as normal text and not as a special
token.

.. code-block:: python

    print(tokenizer.encode("<s>", add_bos=False, add_eos=False))
    # [529, 29879, 29958]

Now let's take a look at LLaMA3's formatting to see how it's tokenized differently
than LLaMA2.

.. code-block:: python

    from torchtune.modules.tokenizers import TikTokenTokenizer

    tokenizer = TikTokenTokenizer("/tmp/Meta-Llama-3-8B/original/tokenizer.model")
    messages = [Message.from_dict(msg) for msg in sample]
    tokens, mask = tokenizer.tokenize_messages(messages)
    print(tokenizer.decode(tokens))
    # '<|start_header_id|>system<|end_header_id|>\n\nYou are a helpful, respectful,
    # and honest assistant.<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nWho
    # are the most influential hip-hop artists of all time?<|eot_id|><|start_header_id|>
    # assistant<|end_header_id|>\n\nHere is a list of some of the most influential hip-hop
    # artists of all time: 2Pac, Rakim, N.W.A., Run-D.M.C., and Nas.<|eot_id|>'

.. note::

>     We used the `tokenize_messages` API for LLaMA3, which is different than

Reviewer: nit: tokenize_messages this is showing up as italics in the doc, think you want double backticks here

    encode. It simply manages adding all the special tokens in the correct
    places after encoding the individual messages.

We can see that the tokenizer handled all the formatting without us specifying a prompt

> template. Turns out, all of the additional tags are special tokens, and we don't require

Reviewer: nit
```suggestion
template. It turns out that all of the additional tags are special tokens, and we don't require
```

a separate prompt template. We can verify this by checking if the tags get encoded
as their own token IDs.

.. code-block:: python

    print(tokenizer._encode_special_token("<|begin_of_text|>"))
    # 128000
    print(tokenizer._encode_special_token("<|eot_id|>"))
    # 128009

The best part is - all these special tokens are handled purely by the tokenizer.
That means you won't have to worry about messing up any required prompt templates!

When should I use a prompt template?
------------------------------------

> When to use or not to use a prompt template is governed by what your desired inference
> behavior is. You should use a prompt template in the following scenarios:

Reviewer: ```suggestion
Whether or not to use a prompt template is governed by your desired inference
behavior. You should use a prompt template in the following scenarios:
```

Author: ah grammar

1. You are running inference on the base model and it was pre-trained with a prompt
template
2. You want to prime a fine-tuned model to expect a certain prompt structure on inference

> for a specific task

Reviewer: This isn't formatting correctly in the live docs

It is not strictly necessary to fine-tune with a prompt template, but generally you
want the model to perform some sort of task, which will require some formatting of

> the prompt.

Reviewer: I might scrap this sentence altogether, esp "you want the model to perform some sort of task" is kinda self-evident

You can fine-tune LLaMA2 with this template even though the model was originally pre-trained
with the :class:`~torchtune.data.Llama2ChatFormat`, as long as this is what the model
sees during inference. The model should be robust enough to adapt to a new template.

Fine-tuning on a custom chat dataset
------------------------------------

Let's test our understanding by trying to fine-tune the LLaMA3-8B model with a custom
chat dataset. We'll walk through how to set up our data so that it can be tokenized
correctly and fed into our model.

Let's say we have a local dataset saved as a CSV file that contains questions
and answers from an online forum. How can we get something like this into a format
LLaMA3 understands and tokenizes correctly?

.. code-block:: python

    import pandas as pd

    df = pd.read_csv('your_file.csv', nrows=1)
    print("Header:", df.columns.tolist())
    # ['input', 'output']
    print("First row:", df.iloc[0].tolist())
    # [
    #     "How do GPS receivers communicate with satellites?",
    #     "The first thing to know is the communication is one-way...",
    # ]

The LLaMA3 tokenizer class, :class:`~torchtune.modules.tokenizers.TikTokenTokenizer`,

> expects the input to be in the :class:`~torchtune.data.Message` format. Let's

Reviewer: You use the `Message` class a few times prior to this. Might be worth a code snippet with a sentence or two explaining everything somewhere in here.

Author: nah, I think the docstrings serve that purpose. Going through eot/ipython/etc would be distracting.

    ) -> ChatDataset:

        return ChatDataset(
            tokenizer=tokenizer,
            # For local csv files, we specify "csv" as the source, just like in
            # load_dataset
            source="csv",
            convert_to_messages=message_converter,
            # Llama3 does not need a chat format
            chat_format=None,
            max_seq_len=max_seq_len,
            # To load a local file we specify it as data_files just like in
            # load_dataset
            data_files="your_file.csv",
        )

.. note::
    You can pass in any keyword argument for :code:`load_dataset` into all our
    Dataset classes and they will honor them. This is useful for common parameters
    such as specifying the data split with :code:`split` or configuration with
    :code:`name`

Now we're ready to start fine-tuning! We'll use the built-in LoRA single device recipe.
Use the :code:`tune cp` command to get a copy of the :code:`8B_lora_single_device.yaml`
config and update it to use your new dataset.

.. code-block:: yaml

    dataset:
      _component_: path.to.my.custom_dataset
      max_seq_len: 2048

Launch the fine-tune!

.. code-block:: bash

>     $ tune run lora_finetune_single_device --config custom_8B_lora_single_device.yaml epochs=15

Reviewer: So many epochs lol
